client/js/widgets/widget.js

```diff
@@ -311,18 +354,80 @@ export class Widget extends StateManaged {
     const routine = this.get(property) !== undefined ? this.get(property) : property;
 
     for(const original of routine) {
-      const a = JSON.parse(JSON.stringify(original));
+      let a = JSON.parse(JSON.stringify(original));
+      if(typeof a == 'object')
+        a = evaluateVariablesRecursively(a);
       var problems = [];
 
       if(this.get('debug')) console.log(`${this.id}: ${JSON.stringify(original)}`);
 
-      if(a.applyVariables) this.applyVariables(a, variables, problems);
-
       if(a.skip) {
         $('#debugButtonOutput').textContent += '\n\n\nOPERATION SKIPPED: \n' + JSON.stringify(a, null, '  ');
         continue;
       }
 
+      if(typeof a == 'string') {
+        const identifier = '(?:[a-zA-Z0-9_-]|\\\\u[0-9a-fA-F]{4})+';
+        const string     = `'((?:[a-zA-Z0-9,.() _-]|\\\\u[0-9a-fA-F]{4})*)'`;
+        const number     = '(-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?)';
+        const variable   = `(\\$\\{[^}]+\\})`;
+        const parameter  = `(null|true|false|\\[\\]|\\{\\}|${number}|${variable}|${string})`;
+
+        const left       = `var (\\$)?(${identifier})(?:\\.(\\$)?(${identifier}))?`;
+        const operation  = `${identifier}|[=+*/%<!>&|-]{1,2}`;
+
+        const regex      = `^${left} += +(?:${parameter}|(?:${parameter} +)?(${operation})(?: +${parameter})?(?: +${parameter})?(?: +${parameter})?)( +//.+)?`;
```

Comment:
Probably shouldn't capture whitespace before comment and empty comment will cause match to fail.
Replace last group of the expression with `(?: +(//.*))?`

Reply:
The capture is not used anywhere.

Do we want empty comments though? Feels like a syntax error to me to write `var a = 1 //`. It's allowed in other languages I guess but what's the point?

Reply:
I certainly wouldn't want to write code that has an empty comment, but I wouldn't want an empty comment to be a compile time error either.